File: src/exec/trash.ts
import fs from 'node:fs'
import os from 'node:os'
import path from 'node:path'
import { renameNoReplace } from '../native/koffi.ts'

export type TrashKey = { filesPath: string; infoPath: string | null; originalPath: string }

function mode0777(st: fs.BigIntStats): number {
  return Number(st.mode) & 0o777
}

function isSafeUserTrashDir(p: string, uid: number): boolean {
  const st = fs.lstatSync(p, { bigint: true })
  if (st.isSymbolicLink() || !st.isDirectory()) return false
  if (Number(st.uid) !== uid) return false
  return mode0777(st) === 0o700
}

function isValidSharedDotTrash(p: string): boolean {
  try {
    const st = fs.lstatSync(p, { bigint: true })
    if (st.isSymbolicLink() || !st.isDirectory()) return false
    return (Number(st.mode) & 0o1000) !== 0
  } catch {
    return false
  }
}

function deepestExisting(p: string): string {
  let cur = p
  while (cur !== '/' && !fs.existsSync(cur)) cur = path.dirname(cur)
  return cur
}

function mkdir0700(p: string): void {
  fs.mkdirSync(p, { mode: 0o700 })
  if (!isSafeUserTrashDir(p, process.getuid?.() ?? 0)) {
    throw new Error(`unsafe trash dir ${p}`)
  }
}

function ensureUserTrashTree(root: string, uid: number): void {
  if (!fs.existsSync(root)) mkdir0700(root)
  else if (!isSafeUserTrashDir(root, uid)) throw new Error(`unsafe trash root ${root}`)
  for (const sub of ['files', 'info']) {
    const d = path.join(root, sub)
    if (!fs.existsSync(d)) mkdir0700(d)
    else if (!isSafeUserTrashDir(d, uid)) throw new Error(`unsafe trash ${d}`)
  }
}

function mountPoint(abs: string): string {
  const srcDev = fs.lstatSync(abs, { bigint: true }).dev
  let cur = path.dirname(abs)
  let last = abs
  while (cur !== last) {
    const st = fs.lstatSync(cur, { bigint: true })
    if (st.dev !== srcDev) return last
    last = cur
    cur = path.dirname(cur)
  }
  return last
}

function percentEncode(s: string): string {
  return encodeURIComponent(s).replace(/'/g, '%27')
}

export function locateTrash(source: string): { root: string; kind: 'home' | 'top' | 'macos' } {
  const uid = process.getuid?.() ?? 0
  if (process.platform === 'darwin') {
    const t = path.join(os.homedir(), '.Trash')
    if (!fs.existsSync(t)) fs.mkdirSync(t, { recursive: true })
    return { root: t, kind: 'macos' }
  }
  const xdg =
    process.env.XDG_DATA_HOME && process.env.XDG_DATA_HOME.length
      ? process.env.XDG_DATA_HOME
      : path.join(os.homedir(), '.local', 'share')
  const ancestor = deepestExisting(xdg)
  const srcDev = fs.lstatSync(source, { bigint: true }).dev
  if (!fs.existsSync(xdg)) {
    const parts = path.relative(ancestor, xdg).split(path.sep).filter(Boolean)
    let cur = ancestor
    for (const part of parts) {
      cur = path.join(cur, part)
      if (!fs.existsSync(cur)) mkdir0700(cur)
    }
  }
  const homeTrashDev = fs.lstatSync(xdg, { bigint: true }).dev
  if (srcDev === homeTrashDev) {
    const root = path.join(xdg, 'Trash')
    ensureUserTrashTree(root, uid)
    return { root, kind: 'home' }
  }
  const mount = mountPoint(source)
  const shared = path.join(mount, '.Trash')
  if (fs.existsSync(shared) && isValidSharedDotTrash(shared)) {
    const root = path.join(shared, String(uid))
    ensureUserTrashTree(root, uid)
    return { root, kind: 'top' }
  }
  const root = path.join(mount, `.Trash-${uid}`)
  ensureUserTrashTree(root, uid)
  return { root, kind: 'top' }
}

export function uniqueTrashName(root: string, base: string, macos: boolean): { filesPath: string; infoPath: string | null } {
  const filesDir = macos ? root : path.join(root, 'files')
  const infoDir = macos ? null : path.join(root, 'info')
  let name = base
  let n = 0
  for (;;) {
    const filesPath = path.join(filesDir, name)
    const infoPath = infoDir ? path.join(infoDir, `${name}.trashinfo`) : null
    const filesFree = !exists(filesPath)
    const infoFree = !infoPath || !exists(infoPath)
    if (filesFree && infoFree) return { filesPath, infoPath }
    n++
    name = `${base}.${n}`
  }
}

function exists(p: string): boolean {
  try {
    fs.lstatSync(p)
    return true
  } catch {
    return false
  }
}

export function writeTrashInfo(opts: {
  infoPath: string
  originalPath: string
  mount: string | null
  kind: 'home' | 'top'
  journalId: string
}): void {
  let pathField: string
  if (opts.kind === 'top' && opts.mount) {
    const rel = opts.originalPath.startsWith(opts.mount + '/')
      ? opts.originalPath.slice(opts.mount.length + 1)
      : opts.originalPath.replace(/^\//, '')
    pathField = percentEncode(rel).replace(/%2F/gi, '/')
  } else {
    pathField = percentEncode(opts.originalPath).replace(/%2F/gi, '/')
  }
  const now = new Date()
  const pad = (n: number) => String(n).padStart(2, '0')
  const deletion = `${now.getFullYear()}-${pad(now.getMonth() + 1)}-${pad(now.getDate())}T${pad(now.getHours())}:${pad(now.getMinutes())}:${pad(now.getSeconds())}`
  const body = `[Trash Info]\nPath=${pathField}\nDeletionDate=${deletion}\nX-Remodeco-Journal=${opts.journalId}\n`
  const fd = fs.openSync(opts.infoPath, 'wx')
  try {
    fs.writeFileSync(fd, body)
    fs.fsyncSync(fd)
  } finally {
    fs.closeSync(fd)
  }
}

export function performTrash(source: string, key: TrashKey, journalId: string, kind: 'home' | 'top' | 'macos'): void {
  if (key.infoPath) {
    const mount = kind === 'top' ? mountPoint(source) : null
    writeTrashInfo({
      infoPath: key.infoPath,
      originalPath: source,
      mount,
      kind: kind === 'macos' ? 'home' : kind,
      journalId,
    })
  }
  renameNoReplace(source, key.filesPath)
}

export function restoreTrash(key: TrashKey): void {
  renameNoReplace(key.filesPath, key.originalPath)
  if (key.infoPath) {
    try {
      fs.unlinkSync(key.infoPath)
    } catch {
      /* ENOENT ok */
    }
  }
}

File: src/trash.rs
use crate::model::TrashKey;
use crate::native::rename_no_replace;
use anyhow::{Context, Result, bail};
use chrono::Local;
use directories::BaseDirs;
use percent_encoding::{AsciiSet, CONTROLS, utf8_percent_encode};
use std::fs::{self, OpenOptions};
use std::io::Write;
use std::os::unix::fs::{MetadataExt, PermissionsExt};
use std::path::{Path, PathBuf};

const PATH_ENCODE: &AsciiSet = &CONTROLS
    .add(b' ')
    .add(b'!')
    .add(b'"')
    .add(b'#')
    .add(b'$')
    .add(b'%')
    .add(b'&')
    .add(b'\'')
    .add(b'(')
    .add(b')')
    .add(b'*')
    .add(b'+')
    .add(b',')
    .add(b':')
    .add(b';')
    .add(b'<')
    .add(b'=')
    .add(b'>')
    .add(b'?')
    .add(b'@')
    .add(b'[')
    .add(b'\\')
    .add(b']')
    .add(b'^')
    .add(b'`')
    .add(b'{')
    .add(b'|')
    .add(b'}')
    .add(b'~');

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum TrashKind {
    Home,
    Top,
    Macos,
}

#[derive(Clone, Debug)]
pub struct TrashLocation {
    pub root: PathBuf,
    pub kind: TrashKind,
}

pub fn locate_trash(source: &Path) -> Result<TrashLocation> {
    let base = BaseDirs::new().context("cannot determine home directory")?;
    #[cfg(target_os = "macos")]
    {
        let root = base.home_dir().join(".Trash");
        fs::create_dir_all(&root)?;
        return Ok(TrashLocation {
            root,
            kind: TrashKind::Macos,
        });
    }
    #[cfg(target_os = "linux")]
    {
        let uid = unsafe { libc::geteuid() };
        let xdg = std::env::var_os("XDG_DATA_HOME")
            .map(PathBuf::from)
            .unwrap_or_else(|| base.home_dir().join(".local/share"));
        create_private_path(&xdg)?;
        let source_dev = fs::symlink_metadata(source)?.dev();
        if fs::symlink_metadata(&xdg)?.dev() == source_dev {
            let root = xdg.join("Trash");
            ensure_user_trash(&root, uid)?;
            return Ok(TrashLocation {
                root,
                kind: TrashKind::Home,
            });
        }
        let mount = mount_point(source)?;
        let shared = mount.join(".Trash");
        if valid_shared_trash(&shared) {
            let root = shared.join(uid.to_string());
            ensure_user_trash(&root, uid)?;
            return Ok(TrashLocation {
                root,
                kind: TrashKind::Top,
            });
        }
        let root = mount.join(format!(".Trash-{uid}"));
        ensure_user_trash(&root, uid)?;
        Ok(TrashLocation {
            root,
            kind: TrashKind::Top,
        })
    }
}

fn create_private_path(path: &Path) -> Result<()> {
    if path.exists() {
        return Ok(());
    }
    let mut missing = Vec::new();
    let mut current = path;
    while !current.exists() {
        missing.push(current.to_path_buf());
        current = current
            .parent()
            .context("trash path has no existing ancestor")?;
    }
    for directory in missing.iter().rev() {
        fs::create_dir(directory)?;
        fs::set_permissions(directory, fs::Permissions::from_mode(0o700))?;
    }
    Ok(())
}

fn safe_user_directory(path: &Path, uid: u32) -> bool {
    fs::symlink_metadata(path).is_ok_and(|metadata| {
        metadata.is_dir()
            && !metadata.file_type().is_symlink()
            && metadata.uid() == uid
            && metadata.mode() & 0o777 == 0o700
    })
}

fn ensure_user_trash(root: &Path, uid: u32) -> Result<()> {
    for directory in [root.to_path_buf(), root.join("files"), root.join("info")] {
        if !directory.exists() {
            fs::create_dir(&directory)?;
            fs::set_permissions(&directory, fs::Permissions::from_mode(0o700))?;
        }
        if !safe_user_directory(&directory, uid) {
            bail!("unsafe trash directory {}", directory.display());
        }
    }
    Ok(())
}

fn valid_shared_trash(path: &Path) -> bool {
    fs::symlink_metadata(path).is_ok_and(|metadata| {
        metadata.is_dir() && !metadata.file_type().is_symlink() && metadata.mode() & 0o1000 != 0
    })
}

fn mount_point(source: &Path) -> Result<PathBuf> {
    let source_dev = fs::symlink_metadata(source)?.dev();
    let mut last = source.to_path_buf();
    let mut current = source.parent().unwrap_or(Path::new("/")).to_path_buf();
    loop {
        if fs::symlink_metadata(&current)?.dev() != source_dev {
            return Ok(last);
        }
        last = current.clone();
        let Some(parent) = current.parent() else {
            return Ok(last);
        };
        if parent == current {
            return Ok(last);
        }
        current = parent.to_path_buf();
    }
}

pub fn unique_trash_key(location: &TrashLocation, source: &Path) -> Result<TrashKey> {
    let base = source
        .file_name()
        .and_then(|v| v.to_str())
        .context("trash source has invalid filename")?;
    let files_dir = if location.kind == TrashKind::Macos {
        location.root.clone()
    } else {
        location.root.join("files")
    };
    let info_dir = (location.kind != TrashKind::Macos).then(|| location.root.join("info"));
    for number in 0_u64.. {
        let name = if number == 0 {
            base.to_owned()
        } else {
            format!("{base}.{number}")
        };
        let files_path = files_dir.join(&name);
        let info_path = info_dir
            .as_ref()
            .map(|directory| directory.join(format!("{name}.trashinfo")));
        if fs::symlink_metadata(&files_path).is_err()
            && info_path
                .as_ref()
                .is_none_or(|path| fs::symlink_metadata(path).is_err())
        {
            return Ok(TrashKey {
                files_path: path_text(&files_path)?,
                info_path: info_path.as_deref().map(path_text).transpose()?,
                original_path: path_text(source)?,
            });
        }
    }
    unreachable!()
}

pub fn perform_trash(
    source: &Path,
    key: &TrashKey,
    journal_id: &str,
    location: &TrashLocation,
) -> Result<()> {
    if let Some(info_path) = &key.info_path {
        let original = path_text(source)?;
        let path_field = if location.kind == TrashKind::Top {
            let mount = mount_point(source)?;
            source
                .strip_prefix(&mount)
                .unwrap_or(source)
                .to_str()
                .context("invalid trash path")?
                .trim_start_matches('/')
                .to_owned()
        } else {
            original
        };
        let body = format!(
            "[Trash Info]\nPath={}\nDeletionDate={}\nX-Remodeco-Journal={}\n",
            utf8_percent_encode(&path_field, PATH_ENCODE),
            Local::now().format("%Y-%m-%dT%H:%M:%S"),
            journal_id,
        );
        let mut file = OpenOptions::new()
            .write(true)
            .create_new(true)
            .open(info_path)?;
        file.write_all(body.as_bytes())?;
        file.sync_all()?;
    }
    if let Err(error) = rename_no_replace(source, Path::new(&key.files_path)) {
        if let Some(info_path) = &key.info_path {
            let _ = fs::remove_file(info_path);
        }
        return Err(error);
    }
    Ok(())
}

pub fn restore_trash(key: &TrashKey) -> Result<()> {
    rename_no_replace(Path::new(&key.files_path), Path::new(&key.original_path))?;
    if let Some(info_path) = &key.info_path {
        match fs::remove_file(info_path) {
            Ok(()) => {}
            Err(error) if error.kind() == std::io::ErrorKind::NotFound => {}
            Err(error) => return Err(error).with_context(|| format!("remove {info_path}")),
        }
    }
    Ok(())
}

fn path_text(path: &Path) -> Result<String> {
    Ok(path.to_str().context("path is not UTF-8")?.to_owned())
}
